### ISSUE TITLE
build error about randombytes.h

### ISSUE THREAD

**Issue by CheyiLin** [zeromq/zeromq4-1#110]
env: CentOS 6.7 x86_64, gcc 4.9.1
branch: `master`
rev: `5aa7bdf`

I got 2 different build errors as detailed below.
Is it possible caused by tweetnacl integration added recently?

### Case 1
build out of source, got error `randombytes.h: No such file or directory`
```bash
$ mkdir build && cd build
$ ../zeromq4-1.git/configure --without-documentation --without-libsodium
...
...
$ make
Making all in doc
make[1]: Entering directory `/tmp/build/doc'
make[1]: Nothing to be done for `all'.
make[1]: Leaving directory `/tmp/build/doc'
make[1]: Entering directory `/tmp/build'
  CXX    src/libzmq_la-address.lo
  CXX    src/libzmq_la-clock.lo
  CXX    src/libzmq_la-ctx.lo
../zeromq4-1.git/src/ctx.cpp:51:25: fatal error: randombytes.h: No such file or directory
 #include "randombytes.h"
                         ^
compilation terminated.
make[1]: *** [src/libzmq_la-ctx.lo] Error 1
make[1]: Leaving directory `/tmp/build'
make: *** [all-recursive] Error 1
```

### Case 2
build in source, got error `./tweetnacl/contrib/randombytes/randombytes.h:14:55: error: ISO C++ 1998 does not support 'long long'`
```bash
$ ./configure --without-documentation --without-libsodium
...
...
$ make
Making all in doc
make[1]: Entering directory `/tmp/zeromq4-1.git/doc'
make[1]: Nothing to be done for `all'.
make[1]: Leaving directory `/tmp/zeromq4-1.git/doc'
make[1]: Entering directory `/tmp/zeromq4-1.git'
  CXX    src/libzmq_la-address.lo
  CXX    src/libzmq_la-clock.lo
  CXX    src/libzmq_la-ctx.lo
In file included from src/ctx.cpp:51:0:
./tweetnacl/contrib/randombytes/randombytes.h:14:55: error: ISO C++ 1998 does not support 'long long' [-Werror=long-long]
 extern void randombytes(unsigned char *,unsigned long long);
                                                       ^
cc1plus: all warnings being treated as errors
make[1]: *** [src/libzmq_la-ctx.lo] Error 1
make[1]: Leaving directory `/tmp/zeromq4-1.git'
make: *** [all-recursive] Error 1
```

**Comment by hintjens** [comment]
@minrk, seems the logic makes it impossible to disable encryption if
libsodium is installed on the system. Do you want to look at this or shall
I?

**Comment by minrk** [comment]
It will probably be a couple of days before I get a chance, so feel free.

**Comment by hintjens** [comment]
OK, can you repeat, for the record, the problems and goals here? Are we
happy using the same strategy as on 4.2, so tweetnacl by default, and
libsodium and no-encryption as options?

**Comment by minrk** [comment]
Problem: Many installations of libzmq lack encryption, even though it's always available from tweetnacl

The important thing for me is that users get encryption by default, since the existence of tweetnacl means it is always available. To me, the most logical behavior would be:

1. nothing specified, try libsodium, fallback on tweetnacl
2. `--with-tweetnacl` skips checking for libsodium and forces tweetnacl
3. `--without-security` disables security altogether with (or some other `--without`)

I wouldn't object to tweetnacl by default and `--with-libsodium`. It's simpler, even if most installs should be doing `--with-libsodium` if it's available.

**Comment by hintjens** [comment]
I've fixed this on zeromq4-1, tested all configurations and I think also fixed the CMake script to work the same way. It's on https://github.com/zeromq/zeromq4-1/pull/111 pending a merge. If someone can give it a whirl, we're good to go...

**Comment by minrk** [comment]
Closed by #111
